bsip-0044.md

```diff
@@ -0,0 +1,292 @@
+
+#### **Business Approach**
+
+Alice begins by generating a distinct `preimage` of her choosing, notes the `preimage length` and calculates the `preimage hash`. She retains the `preimage` in secret, then creates a new HTLC stipulating that the `depositor` account "alice" will transfer `quantity` "100" "bitUSD" `asset` into the `recipient` account "bob" if a `preimage` is presented matching the `preimage hash` before the `timelock threshold` of 10AM tomorrow. Upon consensus validation of the HTLC, the 100 bitUSD `secured assets` are transferred from Alice's `depositor` account into the HTLC where they remain locked by the `preimage hash` and `timelock threshold`. She then shares the resulting `contract identifier` with Bob.
+
+Bob queries the blockchain for the `contract identifier` Alice provided. He examines to ensure it contains his desired `recipient` account, `asset` symbol, asset `quantity`, `preimage length`, and `timelock threshold`. Bob now creates his own HTLC that will deposit `quantity` "10,000" "BTS" `symbol` into the `recipient` account "alice" from `depositor` account "bob", if a `preimage` that generates the `preimage hash` Bob copied from Alice's HTLC before the `timelock threshold` of 5pm today. Upon consensus validation of Bob's HTLC, his 10,000 BTS `secured assets` are transferred from his `depositor` account and "locked" into the contract. He then shares the resulting `contract identifier` with Alice. Notice Bob specified a `timelock threshold` much shorter than Alice defined in her contract. This ensures Bob will have enough time to observe and use the `preimage` Alice will publish to the blockchain next. 
+
+Alice now examines the HTLC Bob created, ensuring the `preimage hash` and `preimage length` both match the original values she used within her contract. She also verifies her desired `recipient` account "alice", the `quantity`, `symbol`, and the `timelock threshold` agree with her intentions. She now uses her `preimage` to "unlock" Bob's contract. Once consensus validation occurs, the HTLC will transfer the `secured assets` 10,000 BTS into her `recipient` account "alice". This reveals the `preimage` on the BitShares blockchain for Bob to use next. NOTE: She must do this before 5PM. Otherwise, Bob may (and should) reclaim the funds in the contract he created.
+
+Bob can now observe the `preimage` Alice used to "unlock" his HTLC, and he will use it to "unlock" her HTLC to receive the 100 bitUSD `secured assets` into his `recipient` account "bob". NOTE: He must do this before 10AM tomorrow. Otherwise, Alice may (and should) reclaim the funds in the contract she created.
+
```

Comment:
One risk I can see is that after Alice broadcast her preimage, Bob would know it immediately and can redeem funds in Alice's contract immediately. However, it's not guaranteed that Alice can redeem Bob's contract successfully, since it's possible that her transaction fails to get into a block E.G. due to chain reorganization (forking) or transaction censoring.

Actually IMHO proposals are more suitable for this use case, because it can guarantee that 2 transfers both succeed or both fail (see https://github.com/bitshares/bitshares-ui/issues/866).

When it's a cross-chain swap, the situation would be even more complicated. To be safe, time-lock of both contracts should be long enough.

Reply:
Agreed that timelock lengths need to be long. The safest way to handle a cross-chain swap from the user's standpoint is:

1) Alice figures out how long it will take for such a transaction to be irreversible on either chain, and use the longest as the baseline.

2) Alice makes her contract timeout the ``baseline*2`` plus a little extra.

3) Bob makes his contract timeout ``baseline``, plus a little extra

Now Alice should wait until Bob's contract is irreversible before exposing her key. This means it is very important that both parties review the contracts carefully, and know the implications of the parameters in each.

There is the risk that Alice exposes her key, and her transaction is rejected. She would have to get a subsequent transaction pushed through before the timeout of that contract. Hence, waiting until the last moment is not a good idea.

Now, there is always the risk of transaction censoring, but I don't think that is only an HTLC problem.